maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/DistributionStageImpl.java

```diff
@@ -102,33 +81,64 @@ public boolean accept(File file) {
-                    Spacelift.task(downloaded, UnzipTool.class).toDir(toExtract).execute().await();
+    private void extractFile(File downloaded, File toExtract, String downloadedPath) {
+        try {
+            if (downloadedPath.endsWith(".zip")) {
+                Spacelift.task(downloaded, UnzipTool.class).toDir(toExtract).execute().await();
+            } else if (downloadedPath.endsWith(".tar.gz")) {
+                Spacelift.task(downloaded, UntarTool.class).gzip(true).toDir(toExtract).execute().await();
+            } else if (downloadedPath.endsWith(".tar.bz2")) {
+                Spacelift.task(downloaded, UntarTool.class).bzip2(true).toDir(toExtract).execute().await();
+            } else {
+                throw new IllegalArgumentException(
+                   "The distribution " + downloaded + " is compressed by unsupported format. "
+                      + "Supported formats are .zip, .tar.gz, .tar.bz2");
+            }
+        } catch (ExecutionException ee) {
+            throw new IllegalStateException(
+               "Something bad happened when the file: " + downloadedPath + " was being extracted. "
+                  + "For more information see the stacktrace", ee);
+        }
+    }
 
-                } else if (downloadedPath.endsWith(".tar.gz")) {
-                    Spacelift.task(downloaded, UntarTool.class).gzip(true).toDir(toExtract).execute().await();
+    private File checkIfItIsAlreadyExtracted(final String downloadedZipMd5hash) {
 
-                } else if (downloadedPath.endsWith(".tar.bz2")) {
-                    Spacelift.task(downloaded, UntarTool.class).bzip2(true).toDir(toExtract).execute().await();
+        File[] dirs = new File(MAVEN_TARGET_DIR).listFiles(new FileFilter() {
 
-                } else {
-                    throw new IllegalArgumentException(
-                            "The distribution " + downloaded + " is compressed by unsupported format. "
-                                    + "Supported formats are .zip, .tar.gz, .tar.bz2");
-                }
-            } catch (ExecutionException ee) {
-                throw new IllegalStateException(
-                        "Something bad happened when the file: " + downloadedPath + " was being extracted. "
-                                + "For more information see the stacktrace");
+            @Override
+            public boolean accept(File file) {
+                return file.isDirectory() && downloadedZipMd5hash.equals(file.getName());
             }
+        });
+        if (dirs != null && containsExtractedFiles(dirs)) {
+            return dirs[0];
+        }
+        return null;
+    }
 
-        return toExtract;
+    private boolean containsExtractedFiles(File[] dirs) {
+        if (dirs.length > 0) {
+            final File[] files = dirs[0].listFiles(new FileFilter() {
+                @Override
+                public boolean accept(File file) {
+                    return file.isDirectory();
```

Comment:
correct me if I'm wrong - this just checks that there is the inner parent directory extracted, not that all inner files and subdirectories in it are extracted as well. How will this ensure that the `bin/mvn` binary is present?

Reply:
Good point @MatousJobanek. Should we only make sure that we have `bin/mvn` extracted? Do we have some marker file or any other way of checking if extraction has been finished?

Reply:
> Should we only make sure that we have bin/mvn extracted? 

It should be figured out what is needed by Maven Invoker when the path to the directory is passed - if it expects `bin/mvn` or something else. 

> Do we have some marker file or any other way of checking if extraction has been finished?

There is nothing so far. The are two ways how to do it:
* counting the size (or a number of inner files) of the directory and checking if the directory grows or not. This can be problematic in cases when the directory is very big or contains some log files where Maven build stores the logs.
* having a file as a marker that the extraction is being processed - when extraction is done, then it would be removed. There is only one problem - when the process is killed by brute force, but as it is in the `target` then it should be removed in the clean phase. And yeah, a shutdown hook for the file removal is a must.

Looking at the possibilities, complexity and possible problems, I would go with the second option - the file marker.